haystack_experimental/components/agents/agent.py

```diff
@@ -0,0 +1,264 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Any, Dict, List, Optional
+
+from haystack import component, default_from_dict, default_to_dict, Pipeline
+from haystack.components.generators.chat.openai import OpenAIChatGenerator
+from haystack.components.joiners import BranchJoiner
+from haystack.components.routers.conditional_router import ConditionalRouter
+from haystack.dataclasses import ChatMessage
+from haystack.lazy_imports import LazyImport
+from haystack.utils import Secret, deserialize_secrets_inplace
+
+from haystack_experimental.components.tools import ToolInvoker
+from haystack_experimental.tools import Tool
+from haystack_experimental.dataclasses.state import State, _schema_from_dict, _schema_to_dict, _validate_schema
+
+with LazyImport(message="Run 'pip install anthropic-haystack' to use Anthropic.") as anthropic_import:
+    from haystack_integrations.components.generators.anthropic.chat.chat_generator import (
+        AnthropicChatGenerator,
+    )
+
+_PROVIDER_GENERATOR_MAPPING = {
+    "openai": OpenAIChatGenerator,
+    "anthropic": AnthropicChatGenerator,
+}
+
+
+@component
+class Agent:
+    """
+    A Haystack component that implements a tool-using agent with provider-agnostic chat model support.
+
+    The component processes messages and executes tools until a exit_condition condition is met.
+    The exit_condition can be triggered either by a direct text response or by invoking a specific designated tool.
+
+    ### Usage example
+    ```python
+    from haystack.dataclasses import ChatMessage
+    from haystack.tools.tool import Tool
+
+    tools = [Tool(name="calculator", description="..."), Tool(name="search", description="...")]
+    input_types = {"search_depth": int}
+    output_types = {"total_tokens": int}
+
+    agent = Agent(
+        model="anthropic:claude-3",
+        generation_kwargs={"temperature": 0.7},
+        tools=tools,
+        exit_condition="search",
+        input_variables=input_types,
+        output_variables=output_types
+    )
+
+    # Run the agent
+    result = agent.run(
+        messages=[ChatMessage.from_user("Find information about Haystack")],
+        search_depth=2
+    )
+
+    assert "messages" in result  # Contains conversation history
+    assert "total_tokens" in result  # Contains tool execution outputs
+    ```
+    """
+
+    def __init__( # pylint: disable=too-many-positional-arguments
+        self,
+        model: str,
```

Comment:
I think we mentioned about directly passing a component here (so a ChatGenerator) instead of a string. This would allow us to drop the `generation_kwargs` and the `api_key`. Also this current version would not support the AzureOpenAIChatGenerator, BedrockChatGenerator, or the GeminiChatGenerator which all require more secrets to be passed than just the api key.

Reply:
I see... It uses the `Component` protocol. In this case, we would need something more specific: a `ChatGenerator` protocol... IDK if we want to include it in this experiment, but is definitively something to explore...

Reply:
That's fair. I think we should use a protocol for this in the version of Agents we bring into main. But I'm fine for this experiment to only work with a subset of ChatGenerators.

Reply:
I'll change it and pass in `Union[OpenAIChatGenerator, AnthropicChatGenerator]` for now. Since it's just a type hint people could presumably also pass other Generators. I only tested with these two though.

Reply:
Serialization and deserialization would only work for those two, but yeah if just using Haystack code then it could also just work with other chat generators.

Reply:
I just wanted to start implementing this but I hit a roadblock. If we pass in the Generator, we need to define tools twice: on the Agent so that they can be passed to the ToolInvoker and when we initialize the ChatGenerator.

I'd say that this wouldn't help simplicity and could confuse users. For yaml definitions it would be very verbose. Since the main point of this abstraction is to make things simpler, I would like to avoid passing in the initialized ChatGenerator.

Reply:
If we want to explore this path, I think it would be legitimate/feasible to expect generators without tools from users and then assign the tools to the generators at runtime...

Reply:
Thanks @anakin87 I took your suggestion and am passing tools at runtime now. It works quite well. I've also updated the example.